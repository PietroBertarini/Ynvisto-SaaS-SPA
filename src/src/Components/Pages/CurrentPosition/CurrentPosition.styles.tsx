import styled from 'styled-components';

export const Title = styled.div`
  color: #fff;
  font-size: 2.8vw;
  font-family: 'Zurich';
  padding-left: 239px;
`;

export const Main = styled.div`
  width: 90vw;
  margin: auto;
  padding: 0 10vw;
  padding-left: 239px;

  .pieChart {
    width: 100%;
    height: 50vh;
  }
`;

export const TotalDiv = styled.div`
  font-family: 'Zurich';
  color: #fff;
  font-size: 2vw;
  padding-left: 5vw;
`;

export const Table = styled.div`
  width: 80%;
  margin: auto;
  display: flex;
  flex-direction: column;
  flex: 1;
  align-items: center;

  table {
    width: 80%;
    border-collapse: collapse;

    th {
      color: #B1B1B1;
      font-size: 1.2vw;
    }
    
    td {
      color: #fff;
      font-size: 1.1vw;
      border-bottom: 1px solid #B1B1B1;      
    }

    th, td {
      font-weight: normal;
      padding: 5px;
      padding-top: 20px;
    }

    td.leftAlign, th.leftAlign { 
      text-align: left;
      padding-left: 2vw;
    }

    td.rightAlign, th.rightAlign { 
      text-align: right; 
      padding-right: 1.5vw;
    }

  }
`;

export const NoCurrentPositionMessage = styled.div`
  color: #fff;
  font-size: 1.8vw;
  font-family: 'Zurich';
  width: 50vw;
  margin: auto;
  padding-top: 10vh;
  text-align: center;
`;
