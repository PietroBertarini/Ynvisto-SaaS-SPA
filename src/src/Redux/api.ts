import axios, { AxiosResponse } from 'axios';

/**
 * Localhost YnvistoGateway api.
 */
const api = axios.create({
  baseURL: process.env.REACT_APP_API_ENDPOINT,
});

// /**
//  * Check if the one error have status 401.
//  * @param err: Caught error
//  */
// const isAuthError = (err: any): boolean => {
//   const axiosError = err as AxiosError;
//   // Whenever the erro is not from axios, we cannot try to repair it
//   if (!axiosError?.response) {
//     return false;
//   }
//
//   // Gets the error code from the axios response
//   const httpStatusCode = axiosError.response?.status;
//
//   // If the error code is Unauthorized, mark the error as Unauthorized error
//   return httpStatusCode === 401;
// };

/**
 * Always that callSafely catch error with status 401, will refresh
 * api token and call one more time.(Just 2 times), if this error occurs
 * again, than throw this error.
 * @param apiCall: Api with one method(get,post...)
 * @param authRetry: boolean to check if error 401 was caught one time.
 */
export function* callSafely<T>(apiCall: Promise<AxiosResponse<any>>, authRetry = false) {
  try {
    return yield apiCall;
  } catch (e) {
    if (!authRetry) {
      // TODO: Refresh token
      // yield callSafely(apiCall, true);
    }
    throw (e);
  }
}

export default api;
