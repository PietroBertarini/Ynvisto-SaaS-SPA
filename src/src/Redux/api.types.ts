export enum ApiErrorCode {
  BrokerNoteImportingError = 'BrokerNoteImportingError',
  BrokerNoteImportingWarning = 'BrokerNoteImportingWarning',
  UnknownError = 'UnknownError',
}

export interface IApiError extends IApiErrorModel {
  httpStatusCode?: number,
}

export interface IApiErrorModel{
  errorMessage: string,
  errorCode: ApiErrorCode
}
