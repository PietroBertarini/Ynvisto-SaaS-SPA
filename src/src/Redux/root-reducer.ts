import { combineReducers } from 'redux';
import { persistReducer, createTransform } from 'redux-persist';

import storage from 'redux-persist/lib/storage';
import topBarReducer from './Components/TopBar/TopBar.reducer';
import { ITopBarState } from './Components/TopBar/TopBar.types';
import { IMyStocksState } from './Services/MyStocks/MyStocks.types';
import myStocksReducer from './Services/MyStocks/MyStocks.reducer';
import taxReportsReducer, { ITaxReportsState } from './Services/TaxReports/TaxReports.reducer';
import { IAuthorizationState } from './Services/Authorization/Authorization.types';
import authorizationReducer from './Services/Authorization/Authorization.reducer';
import irpfReducer, { IIRPFState } from './Services/IRPF/IRPF.reducer';
import User from '../Classes/User/User.class';
//

// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
// @ts-ignore
const SetTransform = createTransform(
  // transform state on its way to being serialized and persisted.
  (inboundState: any, key) => ({ ...inboundState }),
  // transform state being rehydrated
  (outboundState : IAuthorizationState, key) => ({ ...outboundState, user: User.fromStorage(outboundState.user) }),
  // define which reducers this transform gets called for.
  { whitelist: ['authorization'] },
);
// To add a state at persist, just add the reducer inside whitelist.
const persistConfig = {
  key: 'root',
  storage,
  migrate: (state) => Promise.resolve(state),
  transforms: [SetTransform],
  // taxReports: we dont wanna save old taxReports
  blacklist: ['taxReports', 'irpf'],
};
// To store a state into master State, just add your reducer into State Interface.
export interface State {
    topBar: ITopBarState,
    myStocks: IMyStocksState,
    taxReports: ITaxReportsState,
    authorization: IAuthorizationState,
    irpf: IIRPFState,
}
// To store a reducer into appReducer, just add the reducer into combineReducer.
const appReducer = combineReducers({
  topBar: topBarReducer,
  myStocks: myStocksReducer,
  taxReports: taxReportsReducer,
  authorization: authorizationReducer,
  irpf: irpfReducer,
});

const rootReducer = (state, action) => {
  if (action.type === 'USER_LOGOUT') {
    state = undefined;
  }
  return appReducer(state, action);
};

export default persistReducer(persistConfig, rootReducer);
