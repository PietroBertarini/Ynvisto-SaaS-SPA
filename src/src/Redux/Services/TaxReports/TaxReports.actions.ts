import { IApiError } from '../../api.types';
import MonthlyTaxReport from '../../../Classes/MonthlyTaxReport/MonthlyTaxReport.class';

export interface ICurrentPositions {
  ticker: string,
  quantity: number,
  avgCost: number,
  type: string
}

export interface IOperationsHistory {
  /**
   * .NET Date string. i.e: "2020-11-03T00:00:00"
   */
  date: string,
  ticker: string,
  assetType: string,
  operationType: string,
  avgCost: number,
  price: number,
  quantity: number,
  profit: number,
  volume: number,
  isDayTrade: boolean
}

export interface ITaxReport {
  monthlyReports: MonthlyTaxReport[],
  currentPositions: ICurrentPositions[],
  operations: IOperationsHistory[],
}
export enum ETaxReportTypes {
  FETCH_TAX_REPORT_START = 'FETCH_TAX_REPORT_START',
  FETCH_TAX_REPORT_SUCCESS = 'FETCH_TAX_REPORT_SUCCESS',
  FETCH_TAX_REPORT_FAILURE = 'FETCH_TAX_REPORT_FAILURE',
  CLEAN_TAX_DATA = 'CLEAN_TAX_DATA',
  CLEAN_TAX_ERROR = 'CLEAN_TAX_ERROR',
  MARK_DARF_PAID = 'MARK_DARF_PAID',
  MARK_DARF_UNPAID = 'MARK_DARF_UNPAID'
}

export interface IFetchTaxReportsStart {
  type: typeof ETaxReportTypes.FETCH_TAX_REPORT_START
  payload?: string,
}
export interface ICleanTaxError {
  type: typeof ETaxReportTypes.CLEAN_TAX_ERROR
}
export interface ICleanTaxData {
  type: typeof ETaxReportTypes.CLEAN_TAX_DATA
}

export interface IFetchTaxReportsSuccess {
  type: typeof ETaxReportTypes.FETCH_TAX_REPORT_SUCCESS
  payload: ITaxReport,
}

export interface IFetchTaxReportsFailure {
  type: typeof ETaxReportTypes.FETCH_TAX_REPORT_FAILURE
  payload: IApiError,
}

export interface IMarkDarfPaid {
  type: typeof ETaxReportTypes.MARK_DARF_PAID,
  payload: {
    monthIndex: number,
    year: number,
    paymentId: number
  }
}

export interface IMarkDarfUnpaid{
  type: typeof ETaxReportTypes.MARK_DARF_UNPAID,
  payload: {
    monthIndex: number,
    year: number,
  }
}

export type MyTaxReportsActionTypes =
  IFetchTaxReportsStart |
  IFetchTaxReportsSuccess |
  IFetchTaxReportsFailure |
  ICleanTaxData |
  ICleanTaxError |
  IMarkDarfPaid |
  IMarkDarfUnpaid

export const fetchTaxReportsRequest = (spreadsheetId?: string) : IFetchTaxReportsStart => ({
  type: ETaxReportTypes.FETCH_TAX_REPORT_START,
  payload: spreadsheetId,
});
export const taxReportCleanData = () : ICleanTaxData => ({
  type: ETaxReportTypes.CLEAN_TAX_DATA,
});
export const taxReportCleanErrors = () : ICleanTaxError => ({
  type: ETaxReportTypes.CLEAN_TAX_ERROR,
});

export const fetchTaxReportsSuccess = (taxReports: ITaxReport) :
    IFetchTaxReportsSuccess => ({
  type: ETaxReportTypes.FETCH_TAX_REPORT_SUCCESS,
  payload: taxReports,
});

export const fetchTaxReportsFailure = (apiError: IApiError) : IFetchTaxReportsFailure => ({
  type: ETaxReportTypes.FETCH_TAX_REPORT_FAILURE,
  payload: apiError,
});

export const markDarfPaid = (monthIndex: number, year: number, darfPaymentId: number) : IMarkDarfPaid => ({
  type: ETaxReportTypes.MARK_DARF_PAID,
  payload: { monthIndex, year, paymentId: darfPaymentId },
});

export const markDarfUnpaid = (monthIndex: number, year: number) : IMarkDarfUnpaid => ({
  type: ETaxReportTypes.MARK_DARF_UNPAID,
  payload: { monthIndex, year },
});
