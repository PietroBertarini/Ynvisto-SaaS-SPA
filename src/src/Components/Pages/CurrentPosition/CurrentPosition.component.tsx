import React, { useEffect } from 'react';
import { connect, ConnectedProps, useDispatch } from 'react-redux';
import '../../../Theme/sweetalert2/sweetalert2.scss';
import { Redirect } from 'react-router';
import { ResponsivePie } from '@nivo/pie';
import Swal from 'sweetalert2';
import { State } from '../../../Redux/root-reducer';
import {
  fetchTaxReportsRequest,
} from '../../../Redux/Services/TaxReports/TaxReports.actions';
import Spinner from '../../Components/Spinner/spinner.component';
import { multinesErrorToHtml } from '../../../Redux/Services/Api.utils';
import LabelStocks from '../../Components/LabelStocks/LabelStocks.component';
import { extractDataToChart, getTotalPrice } from './CurrentPosition.utils';
import {
  Main, Title, Table, TotalDiv, NoCurrentPositionMessage,
} from './CurrentPosition.styles';

/**
 * Get the stores states that TaxComponents uses and observe.
 * @param state: the stored stated of the Ynvisto.
 */
const mapStateToProps = (state: State) => ({
  taxReport: state.taxReports.taxReport,
  isFetching: state.taxReports.isFetchingReport,
  spreadsheetAdm: state.authorization.spreadsheetAdmin,
  error: state.taxReports.error,
});
const connector = connect(mapStateToProps);
type PropsFromRedux = ConnectedProps<typeof connector>
type Props = PropsFromRedux

function CurrentPosition({
  taxReport, isFetching, error, spreadsheetAdm,
}: Props) {
  const dispatch = useDispatch();

  useEffect(() => {
    if (taxReport === undefined) {
      dispatch(fetchTaxReportsRequest(spreadsheetAdm));
    }
  }, [dispatch, taxReport, spreadsheetAdm]);

  if (isFetching) {
    return <Spinner />;
  }

  if (error?.httpStatusCode === 401) {
    return <Redirect to="/Logout/" />;
  }

  if (error !== undefined) {
    Swal.fire({
      icon: 'error',
      title: 'Um erro ocorreu durante o cálculo da sua posição atual',
      html: multinesErrorToHtml(error.errorMessage),
    });
  }

  const { data, totalValueChart, isShortChart } = extractDataToChart(taxReport?.currentPositions);

  return (
    <>
      <Title>Meus Ativos</Title>
      <Main>

        {taxReport && data && (
          <>
            <TotalDiv>
              <p>
                Total:
                {' '}
                {getTotalPrice(taxReport.currentPositions)
                  .toLocaleString('pt-br', { style: 'currency', currency: 'BRL' })}
              </p>
            </TotalDiv>

            <div className="pieChart">
              <ResponsivePie
                data={data}
                margin={{
                  top: 20, right: 80, bottom: 40, left: 80,
                }}
                innerRadius={0.5}
                padAngle={0.7}
                enableRadialLabels
                enableSliceLabels
                isInteractive
                cornerRadius={3}
                colors={{ datum: 'data.color' }}
                borderWidth={1}
                borderColor={{ from: 'color', modifiers: [['darker', 0.2]] }}
                radialLabelsSkipAngle={10}
                radialLabelsTextColor="#fff"
                radialLabelsLinkColor={{ from: 'color' }}
                radialLabelsLinkStrokeWidth={4}
                radialLabel={(e) => (isShortChart ? (-1 * e.value) : (e.value))
                  .toLocaleString('pt-br', { style: 'currency', currency: 'BRL' })}
                sliceLabelsSkipAngle={10}
                sliceLabelsTextColor="#000"
                sliceLabel={(e) => `${((e.value / totalValueChart) * 100).toFixed(2).replace('.', ',')}%`}
                legends={[
                  {
                    anchor: 'top-left',
                    direction: 'column',
                    itemDirection: 'left-to-right',
                    justify: false,
                    translateX: 100,
                    translateY: -20,
                    itemsSpacing: 10,
                    itemWidth: 100,
                    itemHeight: 25,
                    itemOpacity: 0.80,
                    symbolSize: 20,
                    symbolShape: 'circle',
                    effects: [
                      {
                        on: 'hover',
                        style: {
                          itemOpacity: 1,
                        },
                      },
                    ],
                  },
                ]}
                theme={{
                  fontSize: 17,
                  legends: { text: { fill: '#fff' } },
                }}
              />
            </div>

            <Table>
              <table>
                <thead>
                  <tr>
                    <th className="leftAlign">Ticker</th>
                    <th className="rightAlign">Quantidade</th>
                    <th className="rightAlign">Preço Médio</th>
                    <th className="leftAlign">Tipo</th>
                  </tr>
                </thead>

                <tbody>
                  {taxReport?.currentPositions.map((position) => (
                    <tr key={position.ticker}>
                      <td className="leftAlign">{position.ticker}</td>
                      <td className="rightAlign">{position.quantity}</td>
                      <td className="rightAlign">
                        R$
                        {' '}
                        {(Math.round(position.avgCost * 100) / 100).toFixed(2).replace('.', ',')}
                      </td>
                      <td className="leftAlign"><LabelStocks title={position.type} /></td>
                    </tr>
                  ))}
                </tbody>
              </table>
            </Table>
          </>
        )}
        {!data && (
          <NoCurrentPositionMessage>
            Você não possui ativos no momento.
          </NoCurrentPositionMessage>
        )}
      </Main>
    </>
  );
}
export default connector(CurrentPosition);
