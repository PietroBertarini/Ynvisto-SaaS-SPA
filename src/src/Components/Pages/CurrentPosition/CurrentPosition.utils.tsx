import { ICurrentPositions } from '../../../Redux/Services/TaxReports/TaxReports.actions';

interface IPositionCount {
  value: number;
  count: number;
  name: string;
}

export interface IChartData {
  id: string;
  label: string;
  value: number;
  color: string;
}

export interface IExtractDataToChart {
  data: IChartData[] | null,
  totalValueChart: number,
  isShortChart: boolean,
}

export function extractDataToChart(currentPositions: ICurrentPositions[] | undefined): IExtractDataToChart {
  const positionsCount: IPositionCount[] = [
    {
      count: 0,
      value: 0,
      name: 'Acoes',
    },
    {
      count: 0,
      value: 0,
      name: 'Etf',
    },
    {
      count: 0,
      value: 0,
      name: 'International',
    },
    {
      count: 0,
      value: 0,
      name: 'Options',
    },
    {
      count: 0,
      value: 0,
      name: 'Crypto',
    },
    {
      count: 0,
      value: 0,
      name: 'Fii',
    },
    {
      count: 0,
      value: 0,
      name: 'BDR',
    },
    {
      count: 0,
      value: 0,
      name: 'SubscriptionRight',
    },
  ];

  if (!currentPositions) return { totalValueChart: 0, data: null, isShortChart: false };

  // sets the value and total quantity of an specific type
  currentPositions.forEach((pos) => {
    incrementCounter(positionsCount, pos);
  });

  // Calculates if we have more of short operations than long operations, if so,
  // the graphic will show only the short operations, if not, only long operations
  let short = 0;
  let long = 0;
  positionsCount.forEach((position) => {
    if (position.value < 0) short += 1;
    else if (position.value > 0) long += 1;
  });

  let totalValue = 0;

  // in case we have more short operations, values must be negative, so we must convert
  // them to positive by multiplying by negative one, so it shows on the graphic
  const response = positionsCount
    .filter((res) => ((long >= short) ? (res.value > 0) : (res.value < 0)))

    .map((position) => {
      totalValue += ((long >= short) ? position.value : (position.value * (-1)));

      return {
        id: getPositionLabel(position.name),
        label: getPositionLabel(position.name),
        value: ((long >= short) ? position.value : (position.value * (-1))),
        color: getColor(position.name),
      };
    });

  // if number of operations short === long, long operations prevails
  return {
    data: response,
    totalValueChart: totalValue,
    isShortChart: (!((long >= short))),
  };
}

function incrementCounter(positionsCount: IPositionCount[], pos: ICurrentPositions) {
  positionsCount.map((position) => {
    if (position.name === pos.type) {
      position.count += pos.quantity;
      position.value += pos.avgCost * pos.quantity;
    }
    return 0;
  });
}

export function getTotalPrice(currentPositions: ICurrentPositions[]): number {
  const total = currentPositions.map((current) => current.quantity * current.avgCost);

  return total.reduce((accum, curr) => accum + curr);
}

function getPositionLabel(name: string): string {
  const names = {
    Acoes: 'Ações',
    Etf: 'ETF',
    Options: 'Opções',
    Crypto: 'Crypto',
    Fii: 'FII',
    BDR: 'BDR',
    SubscriptionRight: 'Subscrição',
    International: 'Internacional',
  };

  return names[name];
}

export function getColor(name: string): string {
  const colors = {
    Acoes: '#64C07F',
    Etf: '#F49A59',
    Options: '#EE6868',
    Crypto: '#5FC2FA',
    Fii: '#BFC769',
    BDR: '#5349BD',
    SubscriptionRight: '#D954DC',
    International: '#0075FF',
  };

  return colors[name];
}
