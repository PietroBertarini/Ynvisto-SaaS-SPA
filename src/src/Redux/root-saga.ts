import { all, call } from 'redux-saga/effects';
import { myStocksSagas } from './Services/MyStocks/MyStocks.sagas';
import { taxReportsSagas } from './Services/TaxReports/TaxReports.sagas';
import { authorizationSagas } from './Services/Authorization/Authorization.sagas';

// Add all sagas into All[]
export default function* rootSaga() {
  yield all([call(myStocksSagas), call(taxReportsSagas), call(authorizationSagas)]);
}
