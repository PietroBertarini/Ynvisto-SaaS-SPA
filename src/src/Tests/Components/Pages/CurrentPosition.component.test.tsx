import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import CurrentPosition from '../../../Components/Pages/CurrentPosition/CurrentPosition.component';
import { store } from '../../../Redux/store';
import { fetchTaxReportsSuccess } from '../../../Redux/Services/TaxReports/TaxReports.actions';
import {
  mockCurrentPositionsLong, mockChartDataMoreLongOp, mockChartDataMoreShortOp, mockCurrentPositionsShort,
} from '../../../Mocks/CurrentPosition.mock';
import { extractDataToChart, getTotalPrice } from '../../../Components/Pages/CurrentPosition/CurrentPosition.utils';

export const createMockStore = () => ({
  ...store,
  persistor: {
    persist: () => null,
  },
});

describe('CurrentPosition tests', () => {
  let wrapper: any;
  const mockStore = createMockStore();

  beforeEach(() => {
    mockStore.dispatch(fetchTaxReportsSuccess(mockCurrentPositionsLong));
    mockStore.dispatch = jest.fn();
    wrapper = mount(
      <BrowserRouter>
        <Provider store={mockStore}>
          <CurrentPosition />
        </Provider>
      </BrowserRouter>,
    );
  });
  it('Renders CurrentPosition page and match to previously snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Returns CurrentPosition Chart Data when there are more long operations than short operations', () => {
    const chartData = extractDataToChart(mockCurrentPositionsLong.currentPositions);
    expect(chartData).toEqual(mockChartDataMoreLongOp);
  });

  it('Returns CurrentPosition Chart Data when there are more short operations than long operations', () => {
    const chartData = extractDataToChart(mockCurrentPositionsShort.currentPositions);
    expect(chartData).toEqual(mockChartDataMoreShortOp);
  });

  it('Tests getTotalPrice of current positions', () => {
    const chartData = getTotalPrice(mockCurrentPositionsShort.currentPositions);
    expect(chartData).toBe(-4267.52);
  });
});
