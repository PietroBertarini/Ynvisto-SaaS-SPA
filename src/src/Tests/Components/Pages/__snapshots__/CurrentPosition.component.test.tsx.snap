// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentPosition tests Renders CurrentPosition page and match to previously snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "dispatch": [MockFunction],
          "getState": [Function],
          "persistor": Object {
            "persist": [Function],
          },
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <Connect(CurrentPosition)>
        <CurrentPosition
          dispatch={[MockFunction]}
          isFetching={false}
          taxReport={
            Object {
              "currentPositions": Array [
                Object {
                  "avgCost": 12.22,
                  "quantity": 100,
                  "ticker": "ABEV3",
                  "type": "Acoes",
                },
                Object {
                  "avgCost": 23.9402,
                  "quantity": -100,
                  "ticker": "ITUB4",
                  "type": "Acoes",
                },
                Object {
                  "avgCost": 11.99,
                  "quantity": -100,
                  "ticker": "HFOF11",
                  "type": "Fii",
                },
                Object {
                  "avgCost": 34,
                  "quantity": 100,
                  "ticker": "BRCO11",
                  "type": "Fii",
                },
                Object {
                  "avgCost": 15.035,
                  "quantity": 100,
                  "ticker": "USDT",
                  "type": "Crypto",
                },
              ],
              "monthlyReports": Array [],
              "operations": Array [],
            }
          }
        >
          <styled.div>
            <div
              className="sc-eCApGN esVHXj"
            >
              Meus Ativos
            </div>
          </styled.div>
          <styled.div>
            <div
              className="sc-jSFkmK ejigWk"
            >
              <styled.div>
                <div
                  className="sc-gKAblj cffmaO"
                >
                  <p>
                    Total:
                     
                    R$ 2.532,48
                  </p>
                </div>
              </styled.div>
              <div
                className="pieChart"
              >
                <ResponsivePie
                  borderColor={
                    Object {
                      "from": "color",
                      "modifiers": Array [
                        Array [
                          "darker",
                          0.2,
                        ],
                      ],
                    }
                  }
                  borderWidth={1}
                  colors={
                    Object {
                      "datum": "data.color",
                    }
                  }
                  cornerRadius={3}
                  data={
                    Array [
                      Object {
                        "color": "#5FC2FA",
                        "id": "Crypto",
                        "label": "Crypto",
                        "value": 1503.5,
                      },
                      Object {
                        "color": "#BFC769",
                        "id": "FII",
                        "label": "FII",
                        "value": 2201,
                      },
                    ]
                  }
                  enableRadialLabels={true}
                  enableSliceLabels={true}
                  innerRadius={0.5}
                  isInteractive={true}
                  legends={
                    Array [
                      Object {
                        "anchor": "top-left",
                        "direction": "column",
                        "effects": Array [
                          Object {
                            "on": "hover",
                            "style": Object {
                              "itemOpacity": 1,
                            },
                          },
                        ],
                        "itemDirection": "left-to-right",
                        "itemHeight": 25,
                        "itemOpacity": 0.8,
                        "itemWidth": 100,
                        "itemsSpacing": 10,
                        "justify": false,
                        "symbolShape": "circle",
                        "symbolSize": 20,
                        "translateX": 100,
                        "translateY": -20,
                      },
                    ]
                  }
                  margin={
                    Object {
                      "bottom": 40,
                      "left": 80,
                      "right": 80,
                      "top": 20,
                    }
                  }
                  padAngle={0.7}
                  radialLabel={[Function]}
                  radialLabelsLinkColor={
                    Object {
                      "from": "color",
                    }
                  }
                  radialLabelsLinkStrokeWidth={4}
                  radialLabelsSkipAngle={10}
                  radialLabelsTextColor="#fff"
                  sliceLabel={[Function]}
                  sliceLabelsSkipAngle={10}
                  sliceLabelsTextColor="#000"
                  theme={
                    Object {
                      "fontSize": 17,
                      "legends": Object {
                        "text": Object {
                          "fill": "#fff",
                        },
                      },
                    }
                  }
                >
                  <ResponsiveWrapper>
                    <div
                      style={
                        Object {
                          "height": "100%",
                          "width": "100%",
                        }
                      }
                    />
                  </ResponsiveWrapper>
                </ResponsivePie>
              </div>
              <styled.div>
                <div
                  className="sc-iCoHVE jvDatO"
                >
                  <table>
                    <thead>
                      <tr>
                        <th
                          className="leftAlign"
                        >
                          Ticker
                        </th>
                        <th
                          className="rightAlign"
                        >
                          Quantidade
                        </th>
                        <th
                          className="rightAlign"
                        >
                          Preço Médio
                        </th>
                        <th
                          className="leftAlign"
                        >
                          Tipo
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr
                        key="ABEV3"
                      >
                        <td
                          className="leftAlign"
                        >
                          ABEV3
                        </td>
                        <td
                          className="rightAlign"
                        >
                          100
                        </td>
                        <td
                          className="rightAlign"
                        >
                          R$
                           
                          12,22
                        </td>
                        <td
                          className="leftAlign"
                        >
                          <LabelStocks
                            title="Acoes"
                          >
                            <styled.div>
                              <div
                                className="sc-dlnjPT kjZtbD"
                              >
                                <styled.div
                                  title="Ações"
                                >
                                  <div
                                    className="sc-hKFyIo eamnqb"
                                    title="Ações"
                                  >
                                    Ações
                                  </div>
                                </styled.div>
                              </div>
                            </styled.div>
                          </LabelStocks>
                        </td>
                      </tr>
                      <tr
                        key="ITUB4"
                      >
                        <td
                          className="leftAlign"
                        >
                          ITUB4
                        </td>
                        <td
                          className="rightAlign"
                        >
                          -100
                        </td>
                        <td
                          className="rightAlign"
                        >
                          R$
                           
                          23,94
                        </td>
                        <td
                          className="leftAlign"
                        >
                          <LabelStocks
                            title="Acoes"
                          >
                            <styled.div>
                              <div
                                className="sc-dlnjPT kjZtbD"
                              >
                                <styled.div
                                  title="Ações"
                                >
                                  <div
                                    className="sc-hKFyIo eamnqb"
                                    title="Ações"
                                  >
                                    Ações
                                  </div>
                                </styled.div>
                              </div>
                            </styled.div>
                          </LabelStocks>
                        </td>
                      </tr>
                      <tr
                        key="HFOF11"
                      >
                        <td
                          className="leftAlign"
                        >
                          HFOF11
                        </td>
                        <td
                          className="rightAlign"
                        >
                          -100
                        </td>
                        <td
                          className="rightAlign"
                        >
                          R$
                           
                          11,99
                        </td>
                        <td
                          className="leftAlign"
                        >
                          <LabelStocks
                            title="Fii"
                          >
                            <styled.div>
                              <div
                                className="sc-dlnjPT kjZtbD"
                              >
                                <styled.div
                                  title="FII"
                                >
                                  <div
                                    className="sc-hKFyIo chYOaU"
                                    title="FII"
                                  >
                                    FII
                                  </div>
                                </styled.div>
                              </div>
                            </styled.div>
                          </LabelStocks>
                        </td>
                      </tr>
                      <tr
                        key="BRCO11"
                      >
                        <td
                          className="leftAlign"
                        >
                          BRCO11
                        </td>
                        <td
                          className="rightAlign"
                        >
                          100
                        </td>
                        <td
                          className="rightAlign"
                        >
                          R$
                           
                          34,00
                        </td>
                        <td
                          className="leftAlign"
                        >
                          <LabelStocks
                            title="Fii"
                          >
                            <styled.div>
                              <div
                                className="sc-dlnjPT kjZtbD"
                              >
                                <styled.div
                                  title="FII"
                                >
                                  <div
                                    className="sc-hKFyIo chYOaU"
                                    title="FII"
                                  >
                                    FII
                                  </div>
                                </styled.div>
                              </div>
                            </styled.div>
                          </LabelStocks>
                        </td>
                      </tr>
                      <tr
                        key="USDT"
                      >
                        <td
                          className="leftAlign"
                        >
                          USDT
                        </td>
                        <td
                          className="rightAlign"
                        >
                          100
                        </td>
                        <td
                          className="rightAlign"
                        >
                          R$
                           
                          15,04
                        </td>
                        <td
                          className="leftAlign"
                        >
                          <LabelStocks
                            title="Crypto"
                          >
                            <styled.div>
                              <div
                                className="sc-dlnjPT kjZtbD"
                              >
                                <styled.div
                                  title="Crypto"
                                >
                                  <div
                                    className="sc-hKFyIo bREWOu"
                                    title="Crypto"
                                  >
                                    Crypto
                                  </div>
                                </styled.div>
                              </div>
                            </styled.div>
                          </LabelStocks>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </styled.div>
            </div>
          </styled.div>
        </CurrentPosition>
      </Connect(CurrentPosition)>
    </Provider>
  </Router>
</BrowserRouter>
`;
